submission/hammer/main.go

```diff
@@ -0,0 +1,118 @@
+package main
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+
+	ct "github.com/google/certificate-transparency-go"
+	"github.com/google/certificate-transparency-go/schedule"
+	"github.com/google/certificate-transparency-go/submission"
+)
+
+// Default number of submissions is intentionally low.
+// After several runs, likely to hit Log rate-limits.
+var (
+	proxyEndpoint = flag.String("proxy_endpoint", "http://localhost:5951/", "Endpoint for HTTP (host:port)")
+	timeout       = flag.Duration("duration", 0*time.Minute, "Time to run continuous flow of submissions. "+
+		"When this and count-flag both have non-zero values, submission ends upon reaching earliest restriction")
+	count = flag.Int("count", 10, "Total number of submissions to execute. "+
+		"When this and timeout-flag both have non-zero values, submission ends upon reaching earliest restriction")
+	qps = flag.Int("qps", 5, "Number of requests per second")
+)
+
+func main() {
+	wd, err := os.Getwd()
+	if err != nil {
+		log.Fatal(err)
+	}
+	certData, err := ioutil.ReadFile(filepath.Join(wd, "/submission/hammer/testdata/precert.der"))
+	if err != nil {
+		log.Fatal(err)
+	}
+	interimData, err := ioutil.ReadFile(filepath.Join(wd, "/submission/hammer/testdata/interim.der"))
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	var req ct.AddChainRequest
+	req.Chain = append(req.Chain, certData)
+	req.Chain = append(req.Chain, interimData)
+
+	postBody, err := json.Marshal(req)
+	if err != nil {
+		log.Fatalf("%v\n", err)
+	}
+
+	// Submission runs until timeout or count is reached. If any of those flags
+	// is set to 0, only other flag value is used as restriction.
+	ctx, cancel := context.WithCancel(context.Background())
+	if *timeout > 0 {
+		ctx, cancel = context.WithTimeout(context.Background(), *timeout)
+	} else if *count <= 0 {
+		log.Print("Both timeout and count flags set to 0. No submissions to be sent.")
+		return
+	}
+
+	leftToSend := *count
+	// If count flag is not set, send *qps requests each second until *timeout.
+	if *count <= 0 {
+		leftToSend = int(*timeout / time.Second)
+	}
+	var batchSize int
+
+	schedule.Every(ctx, time.Second, func(ctx context.Context) {
+		batchSize = *qps
+		if leftToSend < *qps {
+			batchSize = leftToSend
+		}
+		leftToSend -= batchSize
+		if batchSize == 0 {
+			cancel()
+		}
+		var wg sync.WaitGroup
+		wg.Add(batchSize)
+		for i := 0; i < batchSize; i++ {
+			go func() {
+				url := *proxyEndpoint + "ct/v1/proxy/add-pre-chain/"
+				resp, err := http.Post(url, "application/json", bytes.NewBuffer(postBody))
+				if err != nil {
+					log.Fatalf("http.Post(%s)=(_,%q); want (_,nil)", url, err)
+				}
+				defer resp.Body.Close()
+				var scts submission.SCTBatch
+				err = json.NewDecoder(resp.Body).Decode(&scts)
+				if err != nil {
+					log.Fatalf("Unable to decode response: %v", err)
```

Comment:
Does `err` make sense without knowing what the response looked like? If it just says there was a problem at character N, for instance, that won't be helpful without knowing what the response actually was.

Reply:
Added into message